tests/test_optimizer.py

```diff
@@ -0,0 +1,224 @@
+            funcs = {}
+            funcs["obj_0"] = 0
+            for x in xdict.keys():
+                funcs["obj_0"] += np.sum(np.power(xdict[x], 2))
+            for iCon, nc in enumerate(self.nCon):
+                conName = f"con_{iCon}"
+                funcs[conName] = np.zeros(nc)
+                for x in xdict.keys():
+                    for j in range(nc):
+                        funcs[conName][j] = (iCon + 1) * np.sum(xdict[x])
+
+            # Throw the fail flag if it's in the specified range or True
+            nonlocal iters
+            if isinstance(failFlag, tuple):
+                if not len(failFlag) == 2:
+                    raise ValueError("Fail flag must be a tuple of (iter start fail, iter end fail) or a boolean")
+                fail = failFlag[0] <= iters < failFlag[1]
+            elif isinstance(failFlag, bool):
+                fail = failFlag
+            else:
+                raise ValueError("Fail flag must be a tuple of (iter start fail, iter end fail) or a boolean")
+            iters += 1
+
+            # Reset iteration counter if any non-finite values in DV dict
+            for xVec in xdict.values():
+                if not np.all(np.isfinite(xVec)):
+                    iters = 0
+                    break
+            return funcs, fail
+
+        return objfunc
+
+    def setup_optProb(self, failFlag=False, nObj=1, nDV=[4], nCon=[2]):
+        """
+        This function sets up a general optimization problem, with arbitrary
+        DVs, constraints and objectives.
+        Arbitrary scaling for the various parameters can also be specified.
+        """
+        self.nObj = nObj
+        self.nDV = nDV
+        self.nCon = nCon
+
+        # Optimization Object
+        self.optProb = Optimization("Configurable Test Problem", self.get_objfunc(failFlag=failFlag))
+        self.x0 = {}
+        # Design Variables
+        for iDV in range(len(nDV)):
+            n = nDV[iDV]
+            x0 = np.ones(n)
+            dvName = f"x{iDV}"
+            self.x0[dvName] = x0
+            self.optProb.addVarGroup(
+                dvName,
+                n,
+                lower=-1,
+                upper=1,
+                value=x0,
+            )
+
+        # Constraints
+        for iCon in range(len(nCon)):
+            nc = nCon[iCon]
+            self.optProb.addConGroup(
+                f"con_{iCon}",
+                nc,
+                lower=-5,
+                upper=5,
+            )
+
+        # Objective
+        for iObj in range(nObj):
+            self.optProb.addObj(f"obj_{iObj}")
+
+        # Finalize
+        self.optProb.printSparsity()
+        # create and store optimizer
+        self.opt = OPT("slsqp", options={"IFILE": "optProb_SLSQP.out"})
+        self.opt(self.optProb, sens="FD")
+
+        # Call the masterFunc with some infinite DVs so it resets iters
+        self.opt._masterFunc(np.full(np.sum(nDV), np.inf), ["fobj"])
+
+    def test_masterFunc_fobj_fail(self):
```

Comment:
Unit tests! Excellent. Do these all fail without the fix in this PR?

Reply:
Yep :) I added a few more for good measure
```
----------------------------------------------------------------------
Ran 10 tests in 0.032s

FAILED (failures=6)
```